File: cmd/modular/resolver.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// ResolutionResult mirrors engine.resolver.ResolutionResult.
type ResolutionResult struct {
	Selected        []*Profile
	Dependencies    map[string][]string
	Packages        []string
	Services        []string
	DisplayProtocol string
}

type Resolver struct {
	reg *Registry
}

func NewResolver(reg *Registry) *Resolver { return &Resolver{reg: reg} }

func qualify(name, prefix string) string {
	if strings.Contains(name, ".") {
		return name
	}
	return prefix + "." + name
}

// Resolve loads requested profiles plus transitive dependencies, merges and
// deduplicates packages/services, detects cycles (spec §32).
func (rs *Resolver) Resolve(desktop string, hardware, applications,
	roles []string) (*ResolutionResult, error) {
	var requested []string
	if desktop != "" && desktop != "none" {
		requested = append(requested, qualify(desktop, "desktop"))
	}
	for _, r := range roles {
		requested = append(requested, qualify(r, "role"))
	}
	for _, h := range hardware {
		requested = append(requested, qualify(h, "hardware"))
	}
	for _, a := range applications {
		requested = append(requested, qualify(a, "app"))
	}

	res := &ResolutionResult{Dependencies: map[string][]string{}}
	resolved := map[string]bool{}
	var visiting []string

	var visit func(id string) error
	visit = func(id string) error {
		if resolved[id] {
			return nil
		}
		for _, v := range visiting {
			if v == id {
				cycle := append(append([]string{}, visiting...), id)
				return fmt.Errorf("dependency cycle detected: %v", cycle)
			}
		}
		p, err := rs.reg.Get(id)
		if err != nil {
			return err
		}
		visiting = append(visiting, id)
		deps := []string{}
		for _, dep := range p.Requires {
			if err := visit(dep); err != nil {
				return err
			}
			deps = append(deps, dep)
		}
		visiting = visiting[:len(visiting)-1]
		resolved[id] = true
		res.Selected = append(res.Selected, p)
		res.Dependencies[id] = deps

		for _, pkg := range p.Packages {
			if !contains(res.Packages, pkg) {
				res.Packages = append(res.Packages, pkg)
			}
		}
		for _, svc := range p.enableServices() {
			if !contains(res.Services, svc) {
				res.Services = append(res.Services, svc)
			}
		}
		if p.Display.Protocol != "" {
			res.DisplayProtocol = p.Display.Protocol
		}
		return nil
	}

	if err := rs.checkConflicts(requested); err != nil {
		return nil, err
	}
	for _, id := range requested {
		if err := visit(id); err != nil {
			return nil, err
		}
	}
	sort.Strings(res.Packages)
	sort.Strings(res.Services)
	return res, nil
}

func (rs *Resolver) checkConflicts(requested []string) error {
	expanded := map[string]bool{}
	for _, id := range requested {
		expanded[id] = true
		p, err := rs.reg.Get(id)
		if err != nil {
			return err
		}
		for _, req := range p.Requires {
			expanded[req] = true
		}
	}
	for _, id := range requested {
		p, _ := rs.reg.Get(id)
		for _, c := range p.Conflicts {
			if contains(requested, c) || (expanded[c] && c != p.ID) {
				return fmt.Errorf("profile conflict: '%s' conflicts with '%s'", p.ID, c)
			}
		}
	}
	return nil
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if x == s {
			return true
		}
	}
	return false
}

func sortProfiles(ps []*Profile) {
	sort.Slice(ps, func(i, j int) bool { return ps[i].ID < ps[j].ID })
}

File: iso/profile/airootfs/opt/modular/engine/resolver.py
"""Dependency resolution engine.

Given a user selection (desktop, hardware features, applications), the
resolver loads the required profiles, pulls in transitive dependencies,
detects conflicts and cycles, and produces a deterministic installation plan.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .errors import ConflictError, DependencyCycleError
from .profiles import Profile, ProfileRegistry


@dataclass
class ResolutionResult:
    selected: list[Profile] = field(default_factory=list)
    dependencies: dict[str, list[str]] = field(default_factory=dict)
    packages: list[str] = field(default_factory=list)
    services: list[str] = field(default_factory=list)
    display_protocol: str | None = None


class Resolver:
    def __init__(self, registry: ProfileRegistry):
        self.registry = registry

    def resolve(self,
                desktop: str | None,
                hardware: list[str] | None = None,
                applications: list[str] | None = None,
                roles: list[str] | None = None) -> ResolutionResult:
        hardware = list(hardware or [])
        applications = list(applications or [])
        roles = list(roles or [])

        requested: list[str] = []
        if desktop and desktop != "none":
            requested.append(self._qualify(desktop, "desktop"))
        for role in roles:
            requested.append(self._qualify(role, "role"))
        for feature in hardware:
            # fully-qualified ids (e.g. gpu.intel, wm.sway) pass through
            requested.append(feature if "." in feature
                             else f"hardware.{feature}")
        for app in applications:
            requested.append(self._qualify(app, "app"))

        result = ResolutionResult()
        resolved: set[str] = set()
        visiting: list[str] = []

        def visit(profile_id: str) -> None:
            if profile_id in resolved:
                return
            if profile_id in visiting:
                cycle = visiting[visiting.index(profile_id):] + [profile_id]
                raise DependencyCycleError(cycle)
            profile = self.registry.get(profile_id)
            visiting.append(profile_id)
            dep_ids: list[str] = []
            for dep in profile.requires:
                visit(dep)
                dep_ids.append(dep)
            visiting.pop()
            resolved.add(profile_id)
            result.selected.append(profile)
            result.dependencies[profile_id] = dep_ids

            for pkg in profile.packages:
                if pkg not in result.packages:
                    result.packages.append(pkg)
            for svc in profile.services.enable:
                if svc not in result.services:
                    result.services.append(svc)
            if profile.display.protocol:
                result.display_protocol = profile.display.protocol

        self._check_conflicts(requested)
        for rid in requested:
            visit(rid)

        result.packages.sort()
        result.services.sort()
        return result

    @staticmethod
    def _qualify(name: str, prefix: str) -> str:
        return name if "." in name else f"{prefix}.{name}"

    def _check_conflicts(self, requested: list[str]) -> None:
        profiles = [self.registry.get(rid) for rid in requested]
        expanded = set(requested)
        for p in profiles:
            expanded.update(p.requires)
        for p in profiles:
            for c in p.conflicts:
                if c in requested or c in expanded - {p.id}:
                    raise ConflictError(p.id, c)
